core/store/src/lib.rs

```diff
@@ -0,0 +1,180 @@
+}
+
+impl Store {
+    pub fn new(storage: Arc<KeyValueDB>) -> Store {
+        Store { storage }
+    }
+
+    pub fn get(&self, column: Option<u32>, key: &[u8]) -> Result<Option<Vec<u8>>, io::Error> {
+        self.storage.get(column, key).map(|a| a.map(|b| b.to_vec()))
+    }
+
+    pub fn get_ser<T: Decode + DeserializeOwned>(
+        &self,
+        column: Option<u32>,
+        key: &[u8],
+    ) -> Result<Option<T>, io::Error> {
+        match self.storage.get(column, key) {
+            Ok(Some(bytes)) => match Decode::decode(bytes.as_ref()) {
+                Ok(result) => Ok(Some(result)),
+                Err(e) => Err(e),
+            },
+            Ok(None) => Ok(None),
+            Err(e) => Err(e),
+        }
+    }
+
+    pub fn exists(&self, column: Option<u32>, key: &[u8]) -> Result<bool, io::Error> {
+        self.storage.get(column, key).map(|value| Option::is_some(&value))
+    }
+
+    pub fn store_update(&self) -> StoreUpdate {
+        StoreUpdate::new(self.storage.clone())
+    }
+}
+
+/// Keeps track of current changes to the database and can commit all of them to the database.
+pub struct StoreUpdate {
+    storage: Arc<KeyValueDB>,
+    transaction: DBTransaction,
+    /// Optionally has reference to the trie to clear cache on the commit.
+    trie: Option<Arc<Trie>>,
+}
+
+impl StoreUpdate {
+    pub fn new(storage: Arc<KeyValueDB>) -> Self {
+        let transaction = storage.transaction();
+        StoreUpdate { storage, transaction, trie: None }
+    }
+
+    pub fn new_with_trie(storage: Arc<KeyValueDB>, trie: Arc<Trie>) -> Self {
+        let transaction = storage.transaction();
+        StoreUpdate { storage, transaction, trie: Some(trie) }
+    }
+
+    pub fn set(&mut self, column: Option<u32>, key: &[u8], value: &[u8]) {
+        self.transaction.put(column, key, value)
+    }
+
+    pub fn set_ser<T: Encode>(
+        &mut self,
+        column: Option<u32>,
+        key: &[u8],
+        value: &T,
+    ) -> Result<(), io::Error> {
+        let data = Encode::encode(value)?;
+        self.set(column, key, &data);
+        Ok(())
+    }
+
+    pub fn delete(&mut self, column: Option<u32>, key: &[u8]) {
+        self.transaction.delete(column, key);
+    }
+
+    /// Merge another store update into this one.
+    pub fn merge(&mut self, other: StoreUpdate) {
+        if self.trie.is_none() {
+            if let Some(trie) = other.trie {
+                self.trie = Some(trie);
+            }
+        }
+        self.merge_transaction(other.transaction);
+    }
+
+    /// Merge DB Transaction.
+    pub fn merge_transaction(&mut self, transaction: DBTransaction) {
+        for op in transaction.ops {
+            match op {
+                DBOp::Insert { col, key, value } => self.transaction.put(col, &key, &value),
+                DBOp::Delete { col, key } => self.transaction.delete(col, &key),
+            }
+        }
+    }
+
+    pub fn commit(self) -> Result<(), io::Error> {
```

Comment:
We need a command opposite to commit, e.g. abort that would purge trie cache and cancel transaction.

Reply:
Will add it separately. Currently we would not have usage for it.